wnfs-common/src/blockstore.rs

```diff
@@ -8,134 +6,181 @@ use libipld::{
+    async fn put_block(&self, bytes: Vec<u8>, codec: IpldCodec) -> Result<Cid> {
+        // Try to build the CID from the bytes and codec
+        let cid = self.create_cid(&bytes, codec)?;
+        // Insert the bytes into the HashMap using the CID as the key
+        self.0.borrow_mut().insert(cid.to_string(), bytes);
+        // Return Ok status with the generated CID
+        Ok(cid)
     }
 }
 
-#[async_trait(?Send)]
-impl BlockStore for MemoryBlockStore {
-    /// Stores an array of bytes in the block store.
-    async fn put_block(&mut self, bytes: Vec<u8>, codec: IpldCodec) -> Result<Cid> {
-        if bytes.len() > MAX_BLOCK_SIZE {
-            bail!(BlockStoreError::MaximumBlockSizeExceeded(bytes.len()))
-        }
+//--------------------------------------------------------------------------------------------------
+// Tests
+//--------------------------------------------------------------------------------------------------
 
-        let hash = Code::Sha2_256.digest(&bytes);
-        let cid = Cid::new(Version::V1, codec.into(), hash)?;
+/// The following methods are generic functions that can be used to test any type that conforms to the BlockStore trait.
+/// In utilizing this structure, externally defined types can still test for retrieval, duplication, and serialization compatibility.
+pub async fn bs_retrieval<T: BlockStore + Send + 'static>(store: &T) -> Result<()> {
+    // Example objects to insert and remove from the blockstore
+    let first_bytes = vec![1, 2, 3, 4, 5];
+    let second_bytes = b"hello world".to_vec();
 
-        self.0.insert(cid.to_string(), bytes);
+    // Insert the objects into the blockstore
+    let first_cid = store.put_serializable(&first_bytes).await.unwrap();
+    let second_cid = store.put_serializable(&second_bytes).await.unwrap();
 
-        Ok(cid)
-    }
+    // Retrieve the objects from the blockstore
+    let first_loaded: Vec<u8> = store.get_deserializable(&first_cid).await.unwrap();
+    let second_loaded: Vec<u8> = store.get_deserializable(&second_cid).await.unwrap();
 
-    /// Retrieves an array of bytes from the block store with given CID.
-    async fn get_block<'a>(&'a self, cid: &Cid) -> Result<Cow<'a, Vec<u8>>> {
-        let bytes = self
-            .0
-            .get(&cid.to_string())
-            .ok_or(BlockStoreError::CIDNotFound(*cid))?;
+    // Assert that the objects are the same as the ones we inserted
+    assert_eq!(first_loaded, first_bytes);
+    assert_eq!(second_loaded, second_bytes);
 
-        Ok(Cow::Borrowed(bytes))
-    }
+    // Return Ok
+    Ok(())
 }
 
-//--------------------------------------------------------------------------------------------------
-// Functions
-//--------------------------------------------------------------------------------------------------
+// Generic function used to test any type that conforms to the BlockStore trait
+pub async fn bs_duplication<T: BlockStore + Send + 'static>(store: &T) -> Result<()> {
```

Comment:
What do you think of this small rename (and similar for the other funcs)?
```suggestion
pub async fn bs_duplication_test<T: BlockStore + Send + 'static>(store: &T) -> Result<()> {
```